src/setup-beam.js

```diff
@@ -278,11 +270,54 @@ async function getRebar3Versions() {
 }
 
 function getVersionFromSpec(spec, versions) {
-  if (versions.includes(spec)) {
-    return spec
+  let version = null
+
+  if (core.getInput('version-type', { required: false }) === 'strict') {
+    version = spec
+  }
+
+  if (version === null) {
+    // We keep a map of semver => "spec" in order to use semver ranges to find appropriate versions
+    const versionsMap = versions.sort(sortVersions).reduce((acc, v) => {
+      try {
+        acc[semver.coerce(v).version] = v
+      } catch {
+        // some stuff can't be coerced, like 'master'
+        acc[v] = v
+      }
+      return acc
+    }, {})
+    const rangeForMax = semver.validRange(spec)
+    if (rangeForMax) {
+      version =
+        versionsMap[semver.maxSatisfying(Object.keys(versionsMap), rangeForMax)]
+    } else {
+      version = versionsMap[spec]
+    }
   }
 
-  return semver.maxSatisfying(versions, semver.validRange(spec))
+  return version
+}
+
+function sortVersions(left, right) {
```

Comment:
This one's a _proper_ attempt at solving OTP-like versions, with 5 elements like 0.0.0.0.0